app/controllers/user_api_key_clients_controller.rb

```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+class UserApiKeyClientsController < ApplicationController
+  layout "no_ember"
+
+  requires_login
+  skip_before_action :check_xhr, :preload_json
+
+  def register
+    require_params
+
+    client = UserApiKeyClient.find_or_initialize_by(client_id: params[:client_id])
+    client.application_name = params[:application_name]
+    client.public_key = params[:public_key]
```

Comment:
Are you planning on adding the `Use stored public_key to encrypt return payload` part in a later commit? Is there a chance we can only store a hashed key for the client, like we do for each key? I'm having a bit of trouble with the inconsistency (why hash in one table but not the other).

Reply:
> Are you planning on adding the Use stored public_key to encrypt return payload part in a later commit? 

That's already done here: https://github.com/discourse/discourse/pull/28119/files#diff-209ab2857786d16d3c404ed2a0dab02c7f4c24105715118407e2803fd42c8760R91

> Is there a chance we can only store a hashed key for the client, like we do for each key?

The key we store in the `user_api_keys` table is the user api key, which should be kept secret, and hence, should be hashed. The `public_key` in the `user_api_key_clients` table is the public key of the client, which does not need to be kept private, and hence doesn't need to be hashed.

Note that previously, i.e. currently, the public key of the client is not stored at all, it's received as a param in the `create` action of the `UserApiKeysController` and then immediately used to encrypt the return payload. As we're now storing the client, it won't actually be necessary for the client to send their public key on every `create` request sent to the `UserApiKeysController`. This is why the public key (and application name) are now only required if there isn't already a stored client:

```
params.require(:public_key) if @client&.public_key.blank?
```